### ISSUE TITLE
[FR] add option to merge received directories with local directories

### ISSUE THREAD

**Issue by NightMachinery** [magic-wormhole/magic-wormhole#454]
E.g., the remote sends `a/` which has two files `a/b/1.txt` and `a/b/2.txt`, but we already have the file `a/b/d.txt` locally. This currently fails because the local directory already exists.

**Comment by piegamesde** [comment]
In practice, merging folders is a lot more complicated than it sounds, especially when you want to get it right and don't want to accidentally overwrite the user's content. Therefore, I'd advise against implementing this. Magic Wormhole is not a tool for syncing folders across machines, after all.

However, I totally agree that simply failing the transfer when the target already exists is frustrating. Having the option to overwrite the target (i.e. without merge) or to receive the content under a different name would be my preferred solutions for this.

**Comment by NightMachinery** [comment]
> In practice, merging folders is a lot more complicated than it sounds, especially when you want to get it right and don't want to accidentally overwrite the user's content. Therefore, I'd advise against implementing this. Magic Wormhole is not a tool for syncing folders across machines, after all.
> 
> However, I totally agree that simply failing the transfer when the target already exists is frustrating. Having the option to overwrite the target (i.e. without merge) or to receive the content under a different name would be my preferred solutions for this.

Can’t we reuse some library for merging directories? Surely some reliable implementation must exist in Python …

**Comment by meejah** [comment]
I agree that "synchronizing folders" isn't _exactly_ a use-case of magic-wormhole (although that said, it does of course fall under the general umbrella of "send files places").

@piegamesde is certainly correct that this isn't as straightforward as it might appear at first. What do you do when a transferred file _exists_ on the target, but is different content? (Among other considerations of course). What if the sender has a _directory_ called `a/foo` but `a/foo` exists as a file on the target?

One approach that might be fruitful would be to take some existing synchronization tool (e.g. `rsync`) and teach it how to use magic-wormhole to set up connections (rather than e.g. `ssh` as it does currently). I think this would probably be far more straightforward on top of Dilation because two-way communication is typically required to get synchronization correct . I don't know `rsync` well enough to know how "pluggable" its transport might be, so this is just a general idea/example.

A key advantage of this would be the potential to re-use the approach for other tools. That is, specifying how to use magic-wormhole as a secure transport. (This may require additional features in magic-wormhole as well as possibly other tools of course). I would personally be a lot more excited to implement it this way instead of creating another `rsync`-like tool

I'm not aware of any Python implementation of general file-synchronizaiont; if you know of any such libraries please post them.

**Comment by NightMachinery** [comment]
My own use case was very simple (the remote only added new files, so no real synchronization needed), but indeed, adding wormhole support to rsync is ideal. 

Perhaps you can go about creating wrappers around rsync that create temporary SSH servers? The wormhole protocol is not in itself useful, the great part is supporting transfer behind NAT and by just a password.

**Comment by piegamesde** [comment]
> One approach that might be fruitful would be to take some existing synchronization tool (e.g. rsync) and teach it how to use magic-wormhole to set up connections (rather than e.g. ssh as it does currently). I think this would probably be far more straightforward on top of Dilation because two-way communication is typically required to get synchronization correct . I don't know rsync well enough to know how "pluggable" its transport might be, so this is just a general idea/example.

I added the feature of forwarding ports to the [Rust version](https://github.com/magic-wormhole/magic-wormhole.rs) for exactly this purpose\*. With that, you can tunnel any application that can work with arbitrary network connections. I think somebody told me they got rsync to work over it, although I haven't tried it out myself yet.

\* The feature is still somewhat experimental, and also broken on the latest release so you'll have to build from source for the time being.

**Comment by meejah** [comment]
Interesting!

Brian and I have talked a little about something similar for Dilation (e.g. to allow other applications to play around with it w/o having to have a Dilation-supporting library for their language).

Thinking out loud, I wonder if SOCKS5 could be an appropriate interface here? It might be stretching the semantics of that protocol a little, but a new connection opening a new sub-channel in Dilation would be the rough idea.

In any case, the point is I'd be a lot more interested in exploring these (port-forwarding, SOCKS5, ...?) as ways to inter-operate with existing applications rather than adding "rsync" features to the Python CLI.

At a high level, the "UX ask" here is reasonable: "synchronize two folders". I just don't think that complexity should live in this code-base. (So, that said, if there _are_ good Python libraries for synchronizing files perhaps this could be a use-case for next-generation file transfer .. that is, the Dilation-based protocol).
